### ISSUE TITLE
Allow multiple installations on the same Windows PC

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#5809]

**Comment by casella** [comment]
Original ticket: https://trac.openmodelica.org/OpenModelica/ticket/5809

**Comment by ceraolo** [comment]
Up to a few weeks ago using multiple installations in Windows was simple.
I created some bat files, one per installation, having content of the following type:

set OPENMODELICAHOME=C:\OM\OpenModelica 1.18.0-64bit\
set OPENMODELICALIBRARY=C:\OM\OpenModelica 1.18.0-64bit\lib\omlibrary
"%OPENMODELICAHOME%"\bin\OMEdit.exe

Unfortunately, this seems not to work anymore with 1.20 and later versions, since Modelica is not loaded.

Is there a similar way to use multiple installations through bat files with newer OM versions?

**Comment by casella** [comment]
@ceraolo, until 1.19.0 we installed libraries in the OpenModelica installation directory, but from 1.20.0 we're using the package manager for that, which installs libraries in the user's .openmodelica directory. This also includes the MSL. So, from1.20.0, the OPENMODELICALIBRARY pathname is no longer specific of the installed version, but rather:

- `%APPDATA%\.openmodelica\libraries` (e.g. C:\Users\Francesco Casella\AppData\Roaming\.openmodelica on my PC) in Windows
- `~/.openmodelica/libraries` (e.g. /home/casella/.openmodelica/libraries on my server) in Linux

Please update your .bat files for versions >= 1.20.0 accordingly.

**Comment by casella** [comment]
See also the User's Guide [here](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/packagemanager.html#the-package-manager).

**Comment by adeas31** [comment]
Maybe just not setting the `OPENMODELICALIBRARY` environment variable in the batch file will solve the issue.

**Comment by ceraolo** [comment]
> @ceraolo, until 1.19.0 we installed libraries in the OpenModelica installation directory, but from 1.20.0 we're using the package manager for that, which installs libraries in the user's .openmodelica directory. 

I knew this. I also read the release notes that give details on this, on the final user point of view.

Just asked if someone already had bat instructions to share.

**Comment by ceraolo** [comment]
> Maybe just not setting the `OPENMODELICALIBRARY` environment variable in the batch file will solve the issue.

Wow, it worked!
This may be useful also for others, since I've frequently read people asking of multiple instances of OpenModelica

**Comment by mahge** [comment]
For what it is worth I am not entirely sure you even need to set OPENMODELICAHOME on installed OpenModelica. We might (?) print a warning if is missing. However, everything should work just fine. At least as far as I can see. We know where everything is based on where the `omc` executable itself (more specifically where the dll/so) resides.  

We seem to set it internally to communicate between different parts of the compiler but that should not require end users to set it explicitly.

**Comment by casella** [comment]
@ceraolo, if you have a set of instructions to manage multiple installations on Windows that are proven to work, please post them here, I can add them to the user's guide.

**Comment by adeas31** [comment]
> For what it is worth I am not entirely sure you even need to set OPENMODELICAHOME on installed OpenModelica. We might (?) print a warning if is missing. However, everything should work just fine. At least as far as I can see. We know where everything is based on where the `omc` executable itself (more specifically where the dll/so) resides.
> 
> We seem to set it internally to communicate between different parts of the compiler but that should not require end users to set it explicitly.

I am not sure about that. We are still setting the `OPENMODELICAHOME` environment variable from the installer. Maybe it is still used. Would be nice to test without it.

**Comment by casella** [comment]
I grep'd OPENMODELICAHOME in the OpenModelica sources, and got 678 occurrences in 272 lines. I guess removing it won't be trivial 😅

**Comment by perost** [comment]
@adrpo changed the handling of `OPENMODELICAHOME` in a6054a57¹. Since then we just get the path from the actual installation directory, and we never actually read `OPENMODELICAHOME` from the environment anymore.

So unless I'm missing something there doesn't seem to be any point for a user to have `OPENMODELICAHOME` set, since the compiler will just ignore it anyway. I don't have it set at least, and I haven't noticed any issues.

¹ The commit also says that `OPENMODELICALIBRARY` isn't read from the environment any longer, but that was later changed.

**Comment by adrpo** [comment]
The OMC and the QT clients (well, except OMOptim) do not need OPENMODELICAHOME.
However, all the other clients that need to find omc.exe, OMPython, OMJulia are using OPENMODELICAHOME to find the installation path. On Linux this doesn't matter as it uses whatever omc is in the path, but on Windows it does matter.

**Comment by adrpo** [comment]
If OPENMODELICALIBRARY is set, we use it, if not, is constructed from the installation directory and the %APPDATA%. The OM installation on Windows doesn't set it anymore.

**Comment by adrpo** [comment]
Basically we could just don't set OPENMODELICAHOME at install time and most of the tools would work fine, but we need to find some better strategy on how to handle OMPython and such.

**Comment by bilderbuchi** [comment]
Does the Windows installer now allow leaving an existing installation in place? 
So far, the first dialog upon executing has been telling me that there is another installation and it needs to be removed, and the only other option is to cancel installation. :confused:

**Comment by adrpo** [comment]
@bilderbuchi I guess that's true, @adeas31 could we remove that check?

**Comment by adeas31** [comment]
> @bilderbuchi I guess that's true, @adeas31 could we remove that check?

Yeah we can allow installation of different versions but not multiple installations of the same version.

**Comment by casella** [comment]
> > @bilderbuchi I guess that's true, @adeas31 could we remove that check?
> 
> Yeah we can allow installation of different versions but not multiple installations of the same version.

That would be great already. For example, one could keep the latest official version and the latest nightly. Sharing the same installed libraries (if we don't change the package manager structure, that is).

It would be nice if one could do that without fancy scripts. I have 3 versions of Dymola installed on my PC and I find it very useful.

**Comment by adrpo** [comment]
> It would be nice if one could do that without fancy scripts. I have 3 versions of Dymola installed on my PC and I find it very useful.

This is already possible for omc and OMEdit, the only clients that would need this set is OMPyton, OMJulia, etc

**Comment by sjoelund** [comment]
I never set any environment variable on my machine because the university IT prevents me from installing OM... You can simply unzip the installer to a directory and it works ;)

**Comment by bilderbuchi** [comment]
> > > Yeah we can allow installation of different versions but not multiple installations of the same version.

> > It would be nice if one could do that without fancy scripts. I have 3 versions of Dymola installed on my PC and I find it very useful.
> 
> This is already possible for omc and OMEdit, the only clients that would need this set is OMPyton, OMJulia, etc

How?
AFAICT, this limitation of the installer is still in place:

> So far, the first dialog upon executing has been telling me that there is another installation and it needs to be removed, and the only other option is to cancel installation. 😕


In the end, (ideally) I would expect to get entries multiple OMEdit versions in the Start menu and choose which one to start.
Knowing that the installer is a zip file, unzipping that somewhere, then manually linking the correct binary from that folder to the appropriate place might work (and be "simple" for an IT-technical person), but is not a nice user-facing flow for obtaining "multiple installations on the same Windows PC", IMHO.

**Comment by adrpo** [comment]
Yes, we need to play with the installer:
- allow multiple installs
- menu items including the version

**Comment by audrey-jardin** [comment]
Hi,
Having an installer letting you having multiple installs would be really useful. 
As industrial end user, it is common to rework on models a few months or even years later and it is a pain to get back to previous OM configurations on which we are sure the models simulate well.

**Comment by audrey-jardin** [comment]
@adrpo , @casella : Hi, could you let us know for which version you plan to have multiple installs "natively"?

**Comment by casella** [comment]
@adrpo, @adeas31 what do you think?

**Comment by adrpo** [comment]
Yes. We should fix this for 1.26.
Basically add the version number to the `OPENMODELICAHOME`: `OPENMODELICAHOME_v1.26.1`.
The OPENMODLICALIBRARY can stay the same for the differnt versions.

All clients that look for `OPENMODELICAHOME` need minor updates, Basically search the entire environment for the existing variables with this prefix and choose the latest (if not specified which one).

**Comment by adeas31** [comment]
> Yes. We should fix this for 1.26. Basically add the version number to the `OPENMODELICAHOME`: `OPENMODELICAHOME_v1.26.1`. The OPENMODLICALIBRARY can stay the same for the differnt versions.
> 
> All clients that look for `OPENMODELICAHOME` need minor updates, Basically search the entire environment for the existing variables with this prefix and choose the latest (if not specified which one).

The problem is not about installing different versions. The main issue is how to use different versions in parallel i.e., can I run two instances of OMEdit from different versions. This seems difficult with the environment variables approach.

**Comment by casella** [comment]
> The problem is not about installing different versions. The main issue is how to use different versions in parallel i.e., can I run two instances of OMEdit from different versions. This seems difficult with the environment variables approach.

I don't think this use case is so interesting. @audrey-jardin is it interesting for you?

The main use case I see (and I think that's the case for others such as @ceraolo) is that you install both the latest official release and the latest nightly build. Then, you open either one. Closing the released version and opening the nightly takes a few seconds, wherease uninstalling the former and installing the latter takes 10-15 minutes on Windows. That's a huge gain.

Of course this means we should check upon starting that there aren't other running instances of OMEdit with a different version tag, and if so abort the starting with an error message like: "Cannot run two different versions of OMEdit simultaneously, please close OMEdit v. xxx".

**Comment by adrpo** [comment]
You can already run 2 different OMEdit instances at the same time. Just don't set `OPENMODELICAHOME` at all.

**Comment by adrpo** [comment]
Variable `OPENMODELICAHOME` is not needed for omc or OMEdit, is needed just for the clients (OMPython, etc) to find the omc binaries.

**Comment by casella** [comment]
> You can already run 2 different OMEdit instances at the same time. Just don't set `OPENMODELICAHOME` at all.

How do you do that?

**Comment by adrpo** [comment]
Remove OPENMODELICAHOME and start OMEdit from the specific folder.

**Comment by cbu-hw** [comment]
> The main use case I see (and I think that's the case for others such as [@ceraolo](https://github.com/ceraolo)) is that you install both the latest official release and the latest nightly build. Then, you open either one. Closing the released version and opening the nightly takes a few seconds, wherease uninstalling the former and installing the latter takes 10-15 minutes on Windows. That's a huge gain.

I agree -- this is about multiple concurrent _installations_, and selecting one of them to run, without uninstall/reinstall.

btw, your PC must be fast (or not slowed down by IT security things) -- for me an uninstall/reinstall cycle takes _easily_ twice that time, half an hour or so.

> You can already run 2 different OMEdit instances at the same time. Just don't set `OPENMODELICAHOME` at all.

The problem occurs before _running_-- how do you _install_ two versions? The current installer always forces you to remove another existing installation.

**Comment by adrpo** [comment]
Ah, yes! We need to fix the installer, to not force deinstallation just offer it.

Right now, one can just remove the OpenModelica key from the registry then install any other OM version:
`REG DELETE "HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Uninstall\OpenModelica"`

**Comment by casella** [comment]
> Remove OPENMODELICAHOME and start OMEdit from the specific folder.

That was my question, how do you remove OPENMODELICAHOME and from where? I need to edit the system path? That looks quite hack-ish.

> Right now, one can just remove the OpenModelica key from the registry then install any other OM version:
> `REG DELETE "HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Uninstall\OpenModelica"`

That's also quite hack-ish, I would not recommend our users to tamper with the registry. We have to support this properly.

**Comment by adrpo** [comment]
The `OPENMODELICAHOME` environment variable you just remove it from the Environment Variables settings. The registry deletion above is for the uninstall information of OM so that you can install another version without uninstalling the current one. I have talked with @adeas31 and we will try to fix this in the near future.

**Comment by adrpo** [comment]
There are several things we need to fix:
- [ ] unique working directory for different OM (OMEdit) versions (include version number or git hash?)
- [ ] different settings for different OM (OMEdit) versions (include version number or git hash?)
- [ ] installer should only propose uninstallation of the current version, not force it
- [ ] do not set the OPENMODELICAHOME environment variable anymore (we need to find another way for clients to find OMC)

**Comment by audrey-jardin** [comment]
> I don't think this use case is so interesting. [@audrey-jardin](https://github.com/audrey-jardin) is it interesting for you?
> 
> The main use case I see (and I think that's the case for others such as [@ceraolo](https://github.com/ceraolo)) is that you install both the latest official release and the latest nightly build. Then, you open either one. Closing the released version and opening the nightly takes a few seconds, wherease uninstalling the former and installing the latter takes 10-15 minutes on Windows. That's a huge gain.
> 
 I agree...it's about having multiple versions of OM without having to uninstall/reinstall or modify the registry.

**Comment by adrpo** [comment]
Check if everything working as it supposed to with the changes:  @adrpo.

**Comment by ceraolo** [comment]
I did a new installation a couple of days ago. It did not accept new installations because some was already installed. 
This is due to the string in regedit:

Computer\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Uninstall\OpenModelica\UninstallString

deleting that one has always worked. I did not make any tests yesterday nor today on Windows.

Probably it is just the matter or removing check of that UnistallString

**Comment by adrpo** [comment]
@adrpo will check this.
